webui/react/src/stores/experiments.tsx

```diff
@@ -0,0 +1,67 @@
+import { Map } from 'immutable';
+import React, { createContext, PropsWithChildren, useCallback, useContext, useState } from 'react';
+
+import { getExperiments } from 'services/api';
+import { V1Pagination } from 'services/api-ts-sdk';
+import { GetExperimentsParams } from 'services/types';
+import { ExperimentItem } from 'types';
+import handleError from 'utils/error';
+import { Loadable, Loaded, NotLoaded } from 'utils/loadable';
+
+type ExperimentPagination = {
+  experiments: ExperimentItem[];
+  pagination?: V1Pagination;
+};
+
+type ExperimentsContext = {
+  experimentsIndex: Map<string, ExperimentPagination>;
+  updateExperimentsIndex: (ws: Map<string, ExperimentPagination>) => void;
+};
+
+const ExperimentsContext = createContext<ExperimentsContext | null>(null);
+
+const encodeArgs = (args: { [key: string]: any }): string => {
+  const orderedKeys = Object.keys(args).sort();
+  return orderedKeys.map((key: string) => `${key}=${JSON.stringify(args[key])}`).join(',');
+};
+
+export const ExperimentsProvider: React.FC<PropsWithChildren> = ({ children }) => {
+  const [experimentsIndex, updateExperimentsIndex] = useState<Map<string, ExperimentPagination>>(
+    Map<string, ExperimentPagination>(),
+  );
+  return (
+    <ExperimentsContext.Provider value={{ experimentsIndex, updateExperimentsIndex }}>
+      {children}
+    </ExperimentsContext.Provider>
+  );
+};
+
+export const useFetchExperiments = (
+  filters: GetExperimentsParams,
+  canceler: AbortController,
+): (() => Promise<void>) => {
+  const context = useContext(ExperimentsContext);
+  if (context === null) {
+    throw new Error('Attempted to use useFetchExperiments outside of Experiment Context');
+  }
+  const { experimentsIndex, updateExperimentsIndex } = context;
+
+  return useCallback(async (): Promise<void> => {
+    try {
+      const response = await getExperiments(filters, { signal: canceler.signal });
+      updateExperimentsIndex(experimentsIndex.set(encodeArgs(filters), response));
+    } catch (e) {
+      handleError(e);
+    }
+  }, [canceler, experimentsIndex, filters, getExperiments, updateExperimentsIndex]);
+};
+
+export const useExperiments = (filters: GetExperimentsParams): Loadable<ExperimentPagination> => {
```

Comment:
it seems like there is an opaque interdependence here. in order to use one of these hooks, you have to also call the other with the exact same args? would prefer to have separate hooks that directly fulfill the requirements. 

in the case of [agents](https://github.com/determined-ai/determined/pull/5358/files), we two separate hooks that fullfilled two different needs "give me the up to date version of this information" and "get me the most recent version of this information that we already have."

Reply:
>While you often want to fetch and use the information in the same place, you don't always. 

so i guess to be exhaustive, in addition to 
- "give me the up to date version of this information"
- and "get me the most recent version of this information that we already have."
we would have
- "refetch this but dont work about giving it to me"
but the latter could be handled by doing "give me the up to date version of this information" and ignoring the return value.


i dont see how having parameters + variations fundamentally changes things. it seems to me like the code as is would pretty much handle all that for us... but maybe i am missing something.

Reply:
I think what you're proposing is fine enough, but would be built on top of this existing code. It would also require you to modify the store code when you have a new usecase. I would encourage us to think of which parts of our code are "library code" and which are "business logic" and try to make the former more stable than the latter.